File: 0778-swim-in-rising-water/0778-swim-in-rising-water.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        dirs = [(0,-1),(0,1),(-1,0),(1,0)]

        def dfs(x, y, t, vis):
            if x == n-1 and y == n-1:
                return True
            vis[x][y] = True
            for dx, dy in dirs:
                nx, ny = x+dx, y+dy
                if 0 <= nx < n and 0 <= ny < n and not vis[nx][ny] and grid[nx][ny] <= t:
                    if dfs(nx, ny, t, vis):
                        return True
            return False

        low, high, ans = 0, n*n - 1, float('inf')
        while low <= high:
            mid = (low + high) // 2
            vis = [[False]*n for _ in range(n)]
            if grid[0][0] <= mid and dfs(0, 0, mid, vis):
                ans = mid
                high = mid - 1
            else:
                low = mid + 1
        return ans

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.java
class Solution {
    int dirs[][] = {{0,-1},{0,1},{-1,0},{1,0}};
    
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        int ans = Integer.MAX_VALUE;
        int low = 0, high = n * n - 1;

        while (low <= high) {
            int mid = (low + high) / 2;
            boolean vis[][] = new boolean[n][n];
            if (grid[0][0] <= mid && dfs(0, 0, mid, grid, vis)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return ans;
    }

    boolean dfs(int x, int y, int t, int[][] grid, boolean[][] vis) {
        int n = grid.length;
        vis[x][y] = true;
        if (x == n - 1 && y == n - 1) return true;

        for (int[] d : dirs) {
            int nx = x + d[0], ny = y + d[1];
            if (nx >= 0 && nx < n && ny >= 0 && ny < n &&
                !vis[nx][ny] && grid[nx][ny] <= t) {
                if (dfs(nx, ny, t, grid, vis)) return true;
            }
        }
        return false;
    }
}